### ISSUE TITLE
Enable feature: Text presets: smart color

### ISSUE THREAD

**Issue by miina** [GoogleForCreators/web-stories-wp#8238]
## Feature Description
The feature "Text presets: smart color" is currently behind a feature flag. For more context: https://github.com/google/web-stories-wp/pull/8118#discussion_r665745831

- [x] Start the image generation when hovering on the library header generally (or at least more widely on the Text tab) instead of directly on the `+` quick-action icon, to get the image ready sooner.
- [x] Do color calculation when hovering on the preset.
- [x] Skip image generation if presets are added one after another (since it's placed below the previous, no need for new image)
- [ ] Add opt-in toggle for using the feature
- [ ] Mark the feature flag `true` by default.

## Additional Context
![image](https://user-images.githubusercontent.com/3294597/134476912-ce76f8a7-85cb-4c81-876c-5eef6539ce3a.png)

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Implementation Brief
- Remove the feature flag (or better -- just mark `true` by default`).
- Do the color calculation already when hovering over the preset, we should already be able to get the information before the click is done.
- If possible, skip image generation if the preset is added below the previous since the area that's being used for calculation is not influenced by the previous text. This would improve the situation only for consecutively added text presets. We would need to move the previously added preset data ref higher up for it being accessible in the places needed.

- Generate the image when hovering anywhere on the text tab of the panel, not just the quick action item specifically.

**Comment by miina** [comment]
@choumx I've created this issue for enabling the feature flag by default for the smart color.

Always disabling the feature for slow devices might be a good idea. If I understood correctly then one idea was checking how long does the image generation take and if it takes more than X seconds, disable the feature. I suppose we should test it out to see how much time does it take in slower devices and add the limit based on that.

We could store this information in `localStorage` for example, however, it might be that the site is just temporarily slow (for other reasons).

Thoughts? Let me know if you had something specific in mind for this.

**Comment by swissspidy** [comment]
Maybe I am missing something, but couldn't this image generation be done async instead of right before element insertion?
Example: I add some elements to the page -> image is created async and stored for later -> then I add a new text preset to the page -> image is already available and insertion is done instantly
This generation could also happen only once I actually go to the Text tab or regularly in a worker.

Sounds less cumbersome than measuring slowness of the device. But anyway just putting some thoughts out there :-)

**Comment by miina** [comment]
> Maybe I am missing something, but couldn't this image generation be done async instead of right before element insertion?
Example: I add some elements to the page -> image is created async and stored for later -> then I add a new text preset to the page -> image is already available and insertion is done instantly
This generation could also happen only once I actually go to the Text tab or regularly in a worker.

The image generation is currently already starting before clicking when:
- [Hovering over the QuickAction icon](https://github.com/google/web-stories-wp/pull/8118/files#diff-0c902e922624e69df6eddb1f2b949f485c7f82b9062fec02d4901e2bd278cd81R153)
- [Hovering over the content of the Text panel tab](https://github.com/google/web-stories-wp/pull/8118/files#diff-7e73c2745fcce1af081c9769d702ca7bdda03707fe85bd5f986308f3648fc2eaR96).

If the image is not ready from before, only then it happens as part of `onClick` (and then it [uses the previously started image generation](https://github.com/google/web-stories-wp/pull/8118/files#diff-b046b35355df967eb1a059d573ecc30f67d9fab305f27c673c07ff5d722b1a67R105) instead of starting it again)

So most of the times it will actually be generated before the user makes a click. We could also change it so that the image would be generated on `pointerEnter` to any part of the library header, and not just the `+` icon of the quick-action as currently. This should improve the situation for the quick-action part a bit.
In case of the text presets, it should already be quite likely the image has been generated by the time user makes a click.

Since the image needs to be generated every time when there is any change happening on the Page then it would cause quite a lot of lagging updating it so frequently, so it seems to make sense to generate it if it looks like there's actually an intention to insert a text.

Thoughts?

**Comment by miina** [comment]
@choumx / @swissspidy Let me know if the optimisations mentioned in the comment above might be enough for enabling the feature for all the users or it still seems better to figure out a way to detect slow devices.

Did a bit more testing with CPU 4x slowdown (and on 3G network). With these settings, the image generation took  on average around 1 second, the image was mostly already ready when getting to the click part. The color calculation took around 30ms.
It looks like the lag comes mainly from the insertion -- not sure what it is exactly since it lags quite noticeable even with just one completely empty page and without the smart color feature. Otherwise I'd guess that re-rendering the page/thumbnail could cause the lag but with an empty page, it would be a bit odd.

Pinging again since if we'll want to get this into the 1.9 release, we'll need it completed quite soon :)
Or should we hold off with this until the text insertion performance has been investigated/improved. Thoughts?

**Comment by dreamofabear** [comment]
Thanks for investigating. 

Re: color calculation, I think 30ms after click is still substantial if our target budget is [50-100ms](https://developer.mozilla.org/en-US/docs/Web/Performance/How_long_is_too_long). 

With some modifications, could color calculation be done at the same time as image generation? E.g. tweak the initial position of inserted text elements (reduce the positional offset for additional text elements) OR sample a larger area of the background for the color calculation. In other words, make text position less variable for color calculation.

Longer-term, users may need to re-trigger this logic (via a quick action?) after moving inserted text to its desired position on the page anyways. So losing some precision in the initial calculation might be acceptable.

> It looks like the lag comes mainly from the insertion -- not sure what it is exactly since it lags quite noticeable even with just one completely empty page and without the smart color feature.

Yea, responsiveness of basic actions like this is an active area of investigation: #8259. I think we should avoid making things worse in the short-term.

@o-fernandez for consideration.

**Comment by o-fernandez** [comment]
Unfortunately, given the additional performance hit for a common use case (adding text) I'd agree with holding the release of this feature until we improve performance.

**Comment by miina** [comment]
Sounds good.

> Re: color calculation, I think 30ms after click is still substantial if our target budget is 50-100ms.

@choumx / @o-fernandez Note that the 30ms result was with slowed down CPU / network. The results with normal settings are more like this, the highest value (6ms) being for the paragraph -- that covers the largest area.
![Screenshot 2021-07-12 at 09 52 13](https://user-images.githubusercontent.com/3294597/125243560-8d8f3a80-e2ee-11eb-8be2-f4f6a001f339.png)

Thoughts?

(Note: this is on the development version which is usually a bit slower than production, too).

> With some modifications, could color calculation be done at the same time as image generation? E.g. tweak the initial position of inserted text elements (reduce the positional offset for additional text elements) OR sample a larger area of the background for the color calculation. In other words, make text position less variable for color calculation.

Hm, we might be able to do the color calculation when hovering over a specific preset button / quick-action button -- we should already be able to determine the position / area depending on which preset is being hovered on.

Just realized there's another optimization we should probably be able do for the image creation for the consecutive text presets -- if we're adding the text preset _below_ the previous one, then we don't really need to re-generate the image. This would improve the performance for that specific case, if the user is adding more than one preset in a row.

I've added these two additional tasks to the description of this issue.


Edit: [I've added a PR](https://github.com/google/web-stories-wp/pull/8285/files) to address these improvements.

**Comment by merapi** [comment]
> couldn't this image generation be done async instead of right before element insertion?

That would be nice if could be done in a non-blocking manner (in a Web Worker). Unfortunately, most things in `html-to-image` require access to the DOM/Canvas.

Another optimization (specific for Smart color) would be to just include the elements from the area that is used for the calculations.

Also, maybe there is a better way to do `getDataURLContent` in `html-to-image` 
![image](https://user-images.githubusercontent.com/21036224/125923579-ebde43c7-3cb4-45e3-96e3-071683b16d72.png)

**Comment by swissspidy** [comment]
> > couldn't this image generation be done async instead of right before element insertion?
> 
> That would be nice if could be done in a non-blocking manner (in a Web Worker). Unfortunately, most things in `html-to-image` require access to the DOM/Canvas.

I was thinking of https://github.com/ampproject/worker-dom there, but I suppose it doesn't cover nearly all of these APIs. 😞 

> Also, maybe there is a better way to do `getDataURLContent` in `html-to-image`
> ![image](https://user-images.githubusercontent.com/21036224/125923579-ebde43c7-3cb4-45e3-96e3-071683b16d72.png)

Impressive how [a one-liner](https://github.com/bubkoo/html-to-image/blob/76f915fe387e8c33d59b44874635b9fc7646e255/src/util.ts#L58-L60) can take so long 😅 

You could try replacing it with something like `dataURL.slice(dataURL.indexOf(',')+1)` which seems faster.

Did some quick comparison with various methods:

Checked test: slice x 7,799,694 ops/sec ±0.34% (67 runs sampled)
Checked test: substr x 7,631,603 ops/sec ±0.74% (65 runs sampled)
Checked test: substring x 7,538,823 ops/sec ±1.33% (66 runs sampled)
Checked test: split x 6,384,898 ops/sec ±0.64% (65 runs sampled)

Could also look into adding some simple memoization.

**Comment by merapi** [comment]
> I was thinking of https://github.com/ampproject/worker-dom there, but I suppose it doesn't cover nearly all of these APIs. 😞

Yep, we can explore that route and see if it makes sense (which parts can be offloaded and what benefit we can get) but without high hopes.

Does that `slice` test include the `indexOf` part?
I was thinking about doing what we need without using that function at all - that would require checking the whole flow of data.

**Comment by merapi** [comment]
> Does that `slice` test include the `indexOf` part?

Those two versions perform roughly the same (it even looks like `split` is marginally faster than `slice+indexOf` for me).

> I was thinking about doing what we need without using that function at all - that would require checking the whole flow of data.

Checked the data flow, but because we can't use Blob on a Canvas directly (only via DataURI b64) there is no way to make less processing here.

**Comment by csossi** [comment]
Please add QA instructions to 8285 - many thanks!

**Comment by barklund** [comment]
Do we have more PR's incoming for this ticket before we can close it since #8285 was only a partial?

**Comment by miina** [comment]
@barklund This part is still not done:
 
>  Mark the feature flag true by default.

@merapi Are there still any pending performance improvement PRs before it can be confirmed by Omar / Will if it's ok to enable the feature? #7985 got merged, but I think you mentioned something else yesterday over the call?

**Comment by merapi** [comment]
I would wait for the #8196 and #7987 to be merged before re-checking the performance, those two fixes should compensate for the canvas image generation (at least partially) and the whole action should feel fast enough.

**Comment by barklund** [comment]
Let's create a new issue just for the feature flag (with a mention of a dependency on good performance) that's blocked by those two other issues, so we can close this one. @merapi?

**Comment by merapi** [comment]
I'm blocking this issue to save time because it's basically what we need "Enable feature, blocked by those 2 issues".

**Comment by dreamofabear** [comment]
Checking in -- our launch criterion should be "does not regress performance".

This means either:
1. Changing smart color logic to avoid adding latency to text insertion. This could be accomplished by making it non-blocking (skip if palette/color is not ready yet), [reducing precision](https://github.com/google/web-stories-wp/issues/8238#issuecomment-877301784), or some other adjustment.
2. Reducing existing text insertion latency to "pay for" the new latency addition. IOW, maintain our current [performance budget](https://web.dev/performance-budgets-101/).

My preference would still be (1) since users will likely move text elements away from the initial inserted position anyways (and eventually add a "magic wand" quick action for text elements to retrigger).

/cc @barklund @o-fernandez

**Comment by dreamofabear** [comment]
From Omar: if skipping is going to be frequent, it may cause user confusion and ultimately may not be worth shipping the feature.

**Comment by barklund** [comment]
This has to be opt-in due to performance hit. @aaskedall to design the required UX for this.

**Comment by miina** [comment]
@barklund I've added design from Andrew to the description and pulled it into this Sprint.
The toggle will work in a consistent way with the toggle of the Page Templates tab.

**Comment by miina** [comment]
Moving this back to UX Needed since there are still pending discussions happening. cc @aaskedall

**Comment by csossi** [comment]
Verified in QA
